src/hardware/imf.cpp

```diff
@@ -0,0 +1,13422 @@
+{
+	return a.note == b.note && a.fraction == b.fraction;
+}
+constexpr bool operator!=(const FractionalNote& a, const FractionalNote& b) noexcept
+{
+	return !(a == b);
+}
+constexpr FractionalNote operator-(const FractionalNote& a,
+                                   const FractionalNote& b) noexcept
+{
+	const auto val = check_cast<uint16_t>(a.getuint16_t() - b.getuint16_t());
+	return to_fractional_note(val);
+}
+
+constexpr FractionalNote operator+(const FractionalNote& a,
+                                   const FractionalNote& b) noexcept
+{
+	// 16-bit wrap-around is expected and normal
+	const auto val = static_cast<uint16_t>(a.getuint16_t() + b.getuint16_t());
+	return to_fractional_note(val);
+}
+
+#pragma pack(push, 1)
+struct KeyVelocity {
+	uint8_t value = 0;
+
+	KeyVelocity() : value(0) {}
+	explicit KeyVelocity(uint8_t v) : value(v) {}
+};
+#pragma pack(pop)
+static_assert(sizeof(KeyVelocity) == 1, "KeyVelocity needs to be 1 in size!");
+
+#pragma pack(push, 1)
+struct Duration {
+	uint16_t value = 0;
+
+	Duration() : value(0) {}
+	explicit Duration(uint16_t v) : value(v) {}
+	inline bool isEmpty() const
+	{
+		return (value & 0x8000) != 0;
+	}
+	inline bool isNotEmpty()
+	{
+		return !isEmpty();
+	}
+	inline void setEmpty()
+	{
+		value |= 0x8000;
+	}
+};
+#pragma pack(pop)
+static_assert(sizeof(Duration) == 2, "Duration2 needs to be 1 in size!");
+
+static Duration ZERO_DURATION(0);
+
+#pragma pack(push, 1)
+struct PitchbenderValueMSB {
+	uint8_t value = 0;
+
+	PitchbenderValueMSB() : value(0) {}
+	explicit PitchbenderValueMSB(uint8_t v) : value(v) {}
+};
+#pragma pack(pop)
+static_assert(sizeof(PitchbenderValueMSB) == 1,
+              "PitchbenderValueMSB needs to be 1 in size!");
+
+#pragma pack(push, 1)
+struct PitchbenderValueLSB {
+	uint8_t value = 0;
+
+	PitchbenderValueLSB() : value(0) {}
+	explicit PitchbenderValueLSB(uint8_t v) : value(v) {}
+};
+#pragma pack(pop)
+static_assert(sizeof(PitchbenderValueLSB) == 1,
+              "PitchbenderValueLSB needs to be 1 in size!");
+
+// clang-format off
+/*
+	Format: <00fedcba>
+	Bit Affected Message                    Status Byte                         Status Byte (Event List)
+	--- ----------------------------------- ----------------------------------- --------------------------
+	a   Note ON/OFF                         [80]~[9F]                           [00]~[2F]
+	b   After Touch and Pitchbend           [A0]~[AF], [D0]~[DF], [E0]~[EF]     [50]~[5F], [60]~[6F]
+	c   Control Change                      [B0]~[BF]                           [30]~[3F]
+		Program Change                      [C0]~[CD]                           [40]~[4F]
+	d   System Exclusive                    [F0], [F7]
+		System Common                       [F1], [F2], [F3], [F6]              [70]~[7F]
+	e   System Real-Time                    [F8], [FA]~[FC]
+	f   ??? only to SP                      N/A                                 N/A
```

Comment:
Wonky indentation

Reply:
I've protected a lot of block comments with `// clang-format off` to prevent it from wrapping and other things. Here's what happens if I take it off here:

![2023-03-28_01-39](https://user-images.githubusercontent.com/1557255/228179624-8c77a69d-da2f-4a36-8222-a130ac4514dd.png)

I've just manually formatted it to the margin now:

```c++

// clang-format off
/*
Format: <00fedcba>
Bit Affected Message                    Status Byte                         Status Byte (Event List)
--- ----------------------------------- ----------------------------------- --------------------------
a   Note ON/OFF                         [80]~[9F]                           [00]~[2F]
b   After Touch and Pitchbend           [A0]~[AF], [D0]~[DF], [E0]~[EF]     [50]~[5F], [60]~[6F]
c   Control Change                      [B0]~[BF]                           [30]~[3F]
	Program Change                      [C0]~[CD]                           [40]~[4F]
d   System Exclusive                    [F0], [F7]
	System Common                       [F1], [F2], [F3], [F6]              [70]~[7F]
e   System Real-Time                    [F8], [FA]~[FC]
f   ??? only to SP                      N/A                                 N/A

Note: Undefined MIDI status bytes ([F4], [F5], [F9], [FD], [FF]) and the active sensing code ([FE]) not listed
      in the preceding table are blocked. Succeeding data bytes are also blocked.
*/
// clang-format on
```